pkg/distributor/distributor.go

```diff
@@ -1005,12 +1042,28 @@ func (d *Distributor) prepareSeriesKeys(ctx context.Context, req *cortexpb.Write
 			continue
 		}
 
+		for _, l := range validation.LimitsPerLabelSetsForSeries(limitsPerLabelSet, cortexpb.FromLabelAdaptersToLabels(validatedSeries.Labels)) {
+			if c, exists := labelSetCounters[l.Hash]; exists {
+				c.floatSamples += int64(len(ts.Samples))
+				c.histogramSamples += int64(len(ts.Histograms))
+			} else {
+				labelSetCounters[l.Hash] = &samplesLabelSetEntry{
+					floatSamples:     int64(len(ts.Samples)),
+					histogramSamples: int64(len(ts.Histograms)),
+					labels:           l.LabelSet,
+				}
+			}
+		}
+
 		seriesKeys = append(seriesKeys, key)
 		validatedTimeseries = append(validatedTimeseries, validatedSeries)
 		validatedFloatSamples += len(ts.Samples)
 		validatedHistogramSamples += len(ts.Histograms)
 		validatedExemplars += len(ts.Exemplars)
 	}
+	for h, counter := range labelSetCounters {
+		d.labelSetCounter.increaseSamplesLabelSet(userID, h, counter.labels, counter.floatSamples, counter.histogramSamples)
+	}
```

Comment:
We decided to buffer and call only once here to minimize the lock contention i guess?

Reply:
Yeah, this way we don't have to lock for each series but instead for each limit